docs/enhancements/patch-install-manifests.md

```diff
@@ -0,0 +1,137 @@
+[`ClusterDeploymentCustomization`](https://github.com/openshift/hive/blob/8c79fedae7011e434e8a7ff0fef40994ef92deb2/vendor/github.com/openshift/hive/apis/hive/v1/clusterdeploymentcustomization_types.go),
+a hive API used by ClusterPools when specific discrete values (such as reserved IP addresses) must
+be injected into the install-config.yaml for each cluster in the pool.
+As such, ClusterDeploymentCustomization originally contained only one member,
+[`InstallConfigPatches`](https://github.com/openshift/hive/blob/8c79fedae7011e434e8a7ff0fef40994ef92deb2/vendor/github.com/openshift/hive/apis/hive/v1/clusterdeploymentcustomization_types.go#L43),
+a list of patches to apply to the install-config.yaml generated by the ClusterPool controller based
+on the user-provided template.
+Each entry in `InstallConfigPatches` is a
+[`PatchEntity`](https://github.com/openshift/hive/blob/8c79fedae7011e434e8a7ff0fef40994ef92deb2/vendor/github.com/openshift/hive/apis/hive/v1/clusterdeploymentcustomization_types.go#L47),
+a hive-owned CRD corresponding to an
+[RFC 6902 JSON patch](https://datatracker.ietf.org/doc/html/rfc6902).
+
+## API
+
+### `ClusterDeploymentCustomization.Spec.InstallerManifestPatches`
+
+**Example:**
+```yaml
+apiVersion: v1
+kind: ClusterDeploymentCustomization
+metadata:
+  name: foo-cluster-deployment-customization
+  namespace: my-project
+spec:
+  installerManifestPatches:
+    # Add custom labels to all master machine manifests
+    - manifestSelector:
+        glob: openshift/99_openshift-cluster-api_master-machines-*.yaml
+      patches:
+        - op: add
+          path: metadata/labels/a-custom-label
+          value: foo
+        - op: add
+          path: metadata/labels/b-custom-label
+          value: foo
+    - ...
+  ...
+```
+
+- Build on the existing [groundwork](#groundwork), extending `ClusterDeploymentCustomization` to
+  expose a new field, `InstallerManifestPatches`.
+- Because installer generates multiple manifests, and RFC 6902 syntax can only reference a single
+  document at a time, we can't directly use `PatchEntity`; we must use an intervening type to
+  identify which file(s) a patch is to be applied to.
+  We'll call this `InstallerManifestPatch`.
+- Each `InstallerManifestPatch`:
+  - Identifies one or more files via subfield `ManifestSelector`.
+  - Lists `PatchEntity`s to apply to files thus identified.
+    This is the same `PatchEntity` used for `InstallConfigPatches`.
+- `ManifestSelector` supports one subfield, `Glob`, which accepts a path matching string as
+  supported by golang's [filepath.Glob](https://pkg.go.dev/path/filepath#Glob).
+  - We will execute the glob relative to the working directory in which manifests were generated.
+  - For security, we will error on any glob starting with `/` or containing `../`.
+- `ManifestSelector` is extensible.
+  E.g. in the future we may wish to match manifests based on the GVK of the object therein.
+
+### `ClusterDeployment.Spec.Provisioning.CustomizationRef`
```

Comment:
Self-

We should fix up the clusterpool controller to automatically add this ref from the inventory to the CD. That way the clusterpool user could do inventory things in either/both ways (install-config and/or manifest patching). Note: requires it not to be a failure if a CDC reffed here contains no manifest patches? Or maybe we can be strict for non-pool clusters? Is it a problem if a non-pool CD uses installConfigPatches??

-Self